### PR TITLE
NN-1240 Key-worker Compliance and Projected Sessions for Individual Staff Member KW

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/KeyworkerStatsController.java [hunk 1]
```diff
@@ -59,4 +61,34 @@ public KeyworkerStatsDto getStatsForStaff(
...
+            nickname="getKeyworkerStats")
+
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "OK", response = PrisonStatsDto.class),
+            @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class ),
+            @ApiResponse(code = 404, message = "Requested resource not found.", response = ErrorResponse.class),
+            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class) })
+
+    @GetMapping(path = "/{prisonId}")
+    public PrisonStatsDto getKeyworkerStats(
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/KeyworkerStatsController.java]: I wonder if a better name would be getStatsForPrison, or getPrisonStats?

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/KeyworkerStatsController.java]: Agreed with change

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java [hunk 2]
```diff
@@ -26,22 +49,81 @@ public KeyworkerStatsDto getStatsForStaff(Long staffId, String prisonId, LocalDa
...
+
+            int projectedKeyworkerSessions = getProjectedKeyworkerSessions(applicableAssignments, staffId, prisonId, fromDate, nextEndDate);
+            final BigDecimal complianceRate = getComplianceRate(sessionsDone, projectedKeyworkerSessions);
+
+            Integer entryCount = usageGroupedBySubType.get(KEYWORKER_ENTRY_SUB_TYPE);
+
+            return KeyworkerStatsDto.builder()
+                    .projectedKeyworkerSessions(projectedKeyworkerSessions)
+                    .complianceRate(complianceRate)
+                    .caseNoteEntryCount(entryCount != null ? entryCount : 0)
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java]: This is done differently to sessionCount (one is inlined the other is not), was this left there by mistake?

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java]: I see, maybe you could make entryCount work like sessionCount?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java [hunk 3]
```diff
@@ -26,22 +49,81 @@ public KeyworkerStatsDto getStatsForStaff(Long staffId, String prisonId, LocalDa
...
+        Validate.notNull(toDate, "toDate");
+
+        return PrisonStatsDto.builder().build();
+    }
+
+    private BigDecimal getComplianceRate(int sessionCount, int projectedKeyworkerSessions) {
+        BigDecimal complianceRate = PERCENT_OPERAND;
+
+        if (projectedKeyworkerSessions > 0)  {
+            complianceRate = new BigDecimal(sessionCount).setScale(4, BigDecimal.ROUND_HALF_UP)
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java]: This looks like a nice API for working calculations. Would using this approach benefit the logic in getProjectedKeyworkerSessions?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java]: Yes I'll tidy this up afterwards. there is alot more to do here for batch so I'll be refactoring this in due cause.

**Pull request merged**